src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Activity.netcoreapp.cs

```diff
@@ -0,0 +1,54 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.Numerics;
+using System.Runtime.CompilerServices;
+using System.Threading;
+
+namespace System.Diagnostics
+{
+    partial class Activity
+    {
+        /// <summary>
+        /// Returns high resolution (~1 usec) current UTC DateTime.
+        /// </summary>
+        internal static DateTime GetUtcNow()
+        {
+            // .NET Core CLR gives accurate UtcNow
+            return DateTime.UtcNow;
+        }
+
+        private unsafe static SpanAction<char, long> s_generateRootId = (span, rootId) =>
+        {
+            span[0] = '|';
+            char format = 'x';
+            span = span.Slice(1);
+            // It is important that the part that changes frequently be first, because
+            // many sampling functions don't sample from the high entropy part of their hash function.
```

Comment:
What is the "high entropy part"? The high bits?

Reply:
Let's say you [use a hash like DJB to get a string hashcode](https://github.com/Microsoft/ApplicationInsights-dotnet/blob/3115fe1cc866a15d09e9b5f1f7f596385406433d/src/ServerTelemetryChannel/Implementation/SamplingScoreGenerator.cs#L45-L50) 
```csharp
int hash = 5381;

for (int i = 0; i < input.Length; i++)
{
    hash = ((hash << 5) + hash) + (int)input[i];
}
```
Each char you add to the hash multiplies the existing hash by 32 adds the original hash, and then adds the char value (likely 32 - 127 for ascii).

The high entropy part is the lowest bits as that's both where the overflow and the new char are getting added. e.g. for "Hello World nnnn"
```
Input String          Hash code  (Hex)
"Hello World 0000" => 1950989825 (0x7449be01)
"Hello World 0001" => 1950989826 (0x7449be02)
"Hello World 0002" => 1950989827 (0x7449be03)
"Hello World 0003" => 1950989828 (0x7449be04)
"Hello World 0004" => 1950989829 (0x7449be05)
```
If you then took that hash value and [divided it by int.MaxValue and multiplied by 100](https://github.com/Microsoft/ApplicationInsights-dotnet/blob/39a5ef23d834777eefdd72149de705a016eb06b0/src/ServerTelemetryChannel/Implementation/SamplingScoreGenerator.cs#L21-L30) to get a percentage-type value you could threshold on that conversion emphasises the low entropy parts of the hash (the high bits)
```
"Hello World 0000" => 1950989825 => 90.8500433856854
"Hello World 0001" => 1950989826 => 90.8500434322516
"Hello World 0002" => 1950989827 => 90.8500434788177
"Hello World 0003" => 1950989828 => 90.8500435253838
"Hello World 0004" => 1950989829 => 90.85004357195
```
Because of this upstream approach to sampling, we front load the part of the string that changes to increase the distribution when its converted to a percentage
```
"0000 Hello World" =>  1568520321 =>  73.0399192185327
"0001 Hello World" => -1394818558 => -64.9513005581458
"0002 Hello World" =>   -63190141 =>  -2.94252024169197
"0003 Hello World" =>  1268438276 =>  59.0662600747618
"0004 Hello World" => -1694900603 => -78.9249597019166
```
It breaks alphanumeric sorting/grouping as the id changes are before server id etc; and I don't think they take into account half the numbers are negative so are only sampling at half rate with thresholding.

However, it produces more variance when only the id is changing for that implementation of sampling.